packages/serve/src/lib/middleware/authMiddleware.ts

```diff
@@ -0,0 +1,79 @@
+import { isEmpty } from 'lodash';
+import { inject, multiInject } from 'inversify';
+import { TYPES as CORE_TYPES, VulcanInternalExtension } from '@vulcan-sql/core';
+import {
+  Next,
+  KoaContext,
+  BuiltInMiddleware,
+  BaseAuthenticator,
+  AuthStatus,
+} from '@vulcan-sql/serve/models';
+import { TYPES } from '@vulcan-sql/serve/containers';
+
+export interface AuthOptions {
+  // different auth type settings
+  [authType: string]: any;
+}
+
+export type AuthenticatorMap = {
+  [name: string]: BaseAuthenticator<any>;
+};
+
+/** The middleware used to check request auth information.
+ *  It seek the 'auth' module name to match data through built-in and customized authenticator by BaseAuthenticator
+ * */
+@VulcanInternalExtension('auth')
+export class AuthMiddleware extends BuiltInMiddleware<AuthOptions> {
+  private authenticators: AuthenticatorMap;
+
+  constructor(
+    @inject(CORE_TYPES.ExtensionConfig) config: any,
+    @inject(CORE_TYPES.ExtensionName) name: string,
+    @multiInject(TYPES.Extension_Authenticator)
+    authenticators: BaseAuthenticator<any>[]
+  ) {
+    super(config, name);
+
+    this.authenticators = authenticators.reduce<AuthenticatorMap>(
+      (prev, authenticator) => {
+        if (authenticator.activate) authenticator.activate();
```

Comment:
Activate functions are asynchronous, we can't wait for them to fulfill in the constructor, maybe we need a activate function for middleware too, and call it from upstream.

Reply:
Thanks for suggestion, I have refactored it, please check :D